src/core/qgsogcutils.cpp

```diff
@@ -1931,15 +1931,31 @@ QDomElement QgsOgcUtils::expressionNodeToOgcFilter( const QgsExpression::Node* n
 
 QDomElement QgsOgcUtils::expressionUnaryOperatorToOgcFilter( const QgsExpression::NodeUnaryOperator* node, QDomDocument& doc, QString& errorMessage )
 {
+
+  QDomElement operandElem = expressionNodeToOgcFilter( node->operand(), doc, errorMessage );
+  if ( !errorMessage.isEmpty() )
+    return QDomElement();
+
   QDomElement uoElem;
   switch ( node->op() )
   {
     case QgsExpression::uoMinus:
       uoElem = doc.createElement( "ogc:Literal" );
-      uoElem.appendChild( doc.createTextNode( "-" ) );
+      if ( node->operand()->nodeType() == QgsExpression::ntLiteral )
+      {
+        // operand expression already created a Literal node:
+        // take the literal value, prepend - and remove old literal node
+        uoElem.appendChild( doc.createTextNode( "-" + operandElem.text() ) );
+        doc.removeChild(operandElem);
+      }
+      else // not sure if this will ever happen
+      {
+        uoElem.appendChild( doc.createTextNode( "-" ) );
```

Comment:
In theory this may happen - e.g. with expression "-fldA + fldB = 0"

From a quick look into OGC Filter Encoding specs it seems to me that v1.1 defined arithmetic operators like <Add> and <Sub>, the latest v2.0 does not mention them anymore. The expression "-X" could be handled the same way as "0-X" is handled - with <Sub> tag.

Or we could just return an error message stating it is not supported - better than just creating invalid filter...